File: problems/problems_827/solution.py
import copy

import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.largestIsland(test_input)

    def largestIsland(self, grid: List[List[int]]) -> int:
        DIRS = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        def point_to_idx(x, y):
            return x * n + y

        n = len(grid)
        uf = UnionFind(n * n)
        for i, row in enumerate(grid):
            for j, val in enumerate(row):
                if val == 1:
                    p = point_to_idx(i, j)
                    for dx, dy in DIRS:
                        if 0 <= (nx := i + dx) < n and 0 <= (ny := j + dy) < n and grid[nx][ny] == 1:
                            uf.union(p, point_to_idx(nx, ny))
        ans = max(uf.size)
        for i, row in enumerate(grid):
            for j, val in enumerate(row):
                if val == 0:
                    tot = 1
                    explored = set()
                    for dx, dy in DIRS:
                        if 0 <= (nx := i + dx) < n and 0 <= (ny := j + dy) < n and grid[nx][ny] == 1:
                            root = uf.find(point_to_idx(nx, ny))
                            if root in explored:
                                continue
                            explored.add(root)
                            tot += uf.size[root]
                    ans = max(ans, tot)

        return ans

class UnionFind:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [1] * n
        self.size = [1] * n
        self.cc = n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]  # 路径压缩
            x = self.parent[x]
        return x

    def union(self, x: int, y: int) -> bool:
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x == root_y:
            return False  # 已经在同一集合

        # 按秩合并
        if self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
            self.size[root_x] += self.size[root_y]
        else:
            self.parent[root_x] = root_y
            if self.rank[root_x] == self.rank[root_y]:
                self.rank[root_y] += 1
            self.size[root_y] += self.size[root_x]
        self.cc -= 1
        return True

    def is_connected(self, x:int, y:int) -> bool:
        return self.find(x) == self.find(y)

File: problems/problems_827/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"

using namespace std;
using json = nlohmann::json;

class UnionFind {
  vector<int> fa;
  vector<int> size;

public:
  int cc;
  explicit UnionFind(int n) : fa(n), size(n, 1), cc(n) {
    for (int i = 0; i < n; i++) {
      fa[i] = i;
    }
  }

  int find(int x) {
    if (fa[x] != x) {
      fa[x] = find(fa[x]);
    }
    return fa[x];
  }

  bool merge(int x, int y) {
    int px = find(x), py = find(y);
    if (px == py) {
      return false;
    }
    fa[px] = py;
    size[py] += size[px];
    cc--;
    return true;
  }

  int get_size(int x) { return size[find(x)]; }
};

constexpr array<array<int, 2>, 4> DIRS({{0, 1}, {1, 0}, {0, -1}, {-1, 0}});

class Solution {
public:
  int largestIsland(const vector<vector<int>> &grid) {
    int n = grid.size();
    auto pointToIdx = [&n](int x, int y) -> int { return x * n + y; };
    UnionFind uf(n * n);
    int ans = 0;
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < n; ++j) {
        if (grid[i][j] == 0) {
          continue;
        }
        int p = pointToIdx(i, j);
        for (const auto &dir : DIRS) {
          int nx = i + dir[0], ny = j + dir[1];
          if (nx < 0 || nx >= n || ny < 0 || ny >= n || grid[nx][ny] != 1) {
            continue;
          }
          uf.merge(p, pointToIdx(nx, ny));
        }
        ans = max(ans, uf.get_size(uf.find(p)));
      }
    }
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < n; ++j) {
        if (grid[i][j] != 0) {
          continue;
        }
        int tot = 1;
        unordered_set<int> explored;
        for (const auto &dir : DIRS) {
          int nx = i + dir[0], ny = j + dir[1];
          if (nx < 0 || nx >= n || ny < 0 || ny >= n || grid[nx][ny] != 1) {
            continue;
          }
          int root = uf.find(pointToIdx(nx, ny));
          if (explored.contains(root)) {
            continue;
          }
          explored.insert(root);
          tot += uf.get_size(root);
        }
        ans = max(ans, tot);
      }
    }
    return ans;
  }
};

json leetcode::qubh::Solve(string input_json_values) {
  vector<string> inputArray;
  size_t pos = input_json_values.find('\n');
  while (pos != string::npos) {
    inputArray.push_back(input_json_values.substr(0, pos));
    input_json_values = input_json_values.substr(pos + 1);
    pos = input_json_values.find('\n');
  }
  inputArray.push_back(input_json_values);

  Solution solution;
  vector<vector<int>> grid = json::parse(inputArray.at(0));
  return solution.largestIsland(grid);
}
